### PR TITLE
#140221857 Integrate social media login 

### PR DESCRIPTION
**Pull request by phabbs**
**What does this PR do?**
Authenticate users through their social media accounts

**Description of the task to be completed**
Allow users to login using their facebook account
Allow users to login using their twitter account
Allow users to login using their google account

**What are the relevant pivotal tracker stories**
#140221857

**Screenshots**
wger_pygo1
wger_pygo_facebook
wger_pygo_facebook_logged
wger_pygo_facebook_logged2
wger_pygo_google_logged
wger_pygo_twitter
wger_pygo_twitter_logged

### CODE REVIEW

File: wger/core/forms.py [hunk 1]
```diff
@@ -39,6 +39,7 @@ class UserLoginForm(AuthenticationForm):
     username = forms.CharField(label=_("Username or email"), max_length=254)
 
 
+
```
Comment (njirap) [inline wger/core/forms.py]: what is the purpose of this blank line?

File: wger/core/templates/template.html [hunk 2]
```diff
@@ -47,6 +47,7 @@
 
 
     {% compress css %}
+    <link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-social/5.1.1/bootstrap-social.css">
```
Comment (njirap) [inline wger/core/templates/template.html]: maybe follow `href` linking conventions that this codebase uses..

File: wger/core/templates/user/login.html [hunk 3]
```diff
@@ -11,8 +11,14 @@
...
       method="post"
       class="form-horizontal">
+
     {% trans 'Login' as submit_text %}
     {% render_form_fields form submit_text %}
+  <p align = "right">
+    <a href="{% url 'social:begin' 'facebook' %}?next={{next}}" class="btn btn-sm btn-social btn-facebook"><span class="fa fa-facebook"></span> Login with Facebook</a>
+    <a href="{% url 'social:begin' 'twitter' %}?next={{next}}" class="btn btn-sm btn-social btn-twitter"><span class="fa fa-twitter"></span> Login with Twitter</a>
+    <a href="{% url 'social:begin' 'google-oauth2' %}?next={{next}}" class="btn btn-sm btn-social btn-google"><span class="fa fa-google"></span> Login with Google</a>
+  </p>
```
Comment (njirap) [inline wger/core/templates/user/login.html]: maybe follow indenting conventions..

File: wger/core/views/user.py [hunk 4]
```diff
@@ -72,6 +72,7 @@ def login(request):
     Small wrapper around the django login view
     '''
 
+
```
Comment (njirap) [inline wger/core/views/user.py]: what is the purpose of this blank line?

File: wger/core/views/user.py [hunk 5]
```diff
@@ -16,7 +16,7 @@
 
 import logging
 
-from django.shortcuts import render, get_object_or_404
+from django.shortcuts import render_to_response, redirect, render, get_object_or_404
```
Comment (njirap) [inline wger/core/views/user.py]: why are these imported?

File: wger/settings_global.py [hunk 6]
```diff
@@ -35,13 +39,28 @@
...
+SOCIAL_AUTH_FACEBOOK_KEY = dotenv.get("SOCIAL_AUTH_FACEBOOK_KEY")
+SOCIAL_AUTH_FACEBOOK_SECRET = dotenv.get("SOCIAL_AUTH_FACEBOOK_SECRET")
+SOCIAL_AUTH_GOOGLE_OAUTH2_KEY = dotenv.get("SOCIAL_AUTH_GOOGLE_OAUTH2_KEY")
+SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET = dotenv.get("SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET")
+
+SOCIAL_AUTH_LOGIN_REDIRECT_URL = '/dashboard'
+SOCIAL_AUTH_LOGIN_URL = '/'
+
+
+
```
Comment (njirap) [inline wger/settings_global.py]: too many blank lines here..

File: wger/settings_global.py [hunk 7]
```diff
@@ -16,6 +16,9 @@
 
 import re
 import sys
+import dotenv
+dotenv.load()
```
Comment (njirap) [inline wger/settings_global.py]: why is this operation being done among the imports?

File: requirements.txt [hunk 8]
```diff
@@ -1,30 +1,56 @@
...
+rcssmin==1.0.6
+reportlab==3.3.0
+requests==2.13.0
+requests-oauthlib==0.8.0
+rjsmin==1.0.12
+six==1.10.0
+snowballstemmer==1.2.1
+social-auth-app-django==1.1.0
+social-auth-core==1.2.0
+Sphinx==1.5.2
```
Comment (njirap) [inline requirements.txt]: this requirements file is bloated...let's strive towards a lighter one..

File: wger/utils/middleware.py [hunk 9]
```diff
@@ -119,3 +122,11 @@ def process_response(self, request, response):
             response['X-wger-redirect'] = request.path
             response.content = request.path
         return response
+
+
+class SocialAuthExceptionMiddleware(SocialAuthExceptionMiddleware):
```
Comment (koyagabriel) [inline wger/utils/middleware.py]: what is going on here? Why is a subclass having the same name with its superclass?

File: wger/utils/middleware.py [hunk 10]
```diff
@@ -119,3 +122,11 @@ def process_response(self, request, response):
...
+    def process_exception(self, request, exception):
```
Comment (koyagabriel) [inline wger/utils/middleware.py]: why is request required as an argument in this method?

Comment (phabbs) [reply wger/utils/middleware.py]: The process_exception() method is an in-built function in SocialAuthExceptionMiddleware which is been overwritten by our custom function. Any method calling the function will be passing the request and the exception arguments but if a request is not passed as argument, a positional argument exception will be thrown.
https://github.com/python-social-auth/social-app-django/blob/master/social_django/middleware.py

**Pull request merged**